### ISSUE TITLE
crossgen2 fails with 139 exit code on Arm64

### ISSUE THREAD

**Issue by mthalman** [dotnet/source-build#4007]
This is causing a failure in the [VMR](https://github.com/dotnet/dotnet) for .NET 9 Preview 1 when building the installer repo. It fails in the Ubuntu2204Arm64_Offline_MsftSdk_arm64 job ([example build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2359821&view=logs&s=18955158-b0fd-513e-a201-fa7e2cb94941&j=48f70f0d-6e60-5f3f-20b6-a737f6b57504) [internal link]). 

Error:

```
/vmr/src/installer/src/redist/targets/Crossgen.targets(163,5): error MSB6006: "crossgen2" exited with code 139. [/vmr/src/installer/src/redist/redist.csproj]
```

Here is a binlog:
[sourcebuild.installer.zip](https://github.com/dotnet/source-build/files/14030433/sourcebuild.installer.zip)

We need this resolved for .NET 9 Preview 1.

**Comment by mthalman** [comment]
@jkoritzinsky - Can you help investigate? This was working recently and regressed.

**Comment by jkoritzinsky** [comment]
I can try but I don't have a Linux ARM64 machine. Do you know the first VMR build that this broke in? I'll see if I can narrow down where the failure from.

**Comment by jkoritzinsky** [comment]
It looks like the most recent passing build was https://dev.azure.com/dnceng/internal/_build/results?buildId=2359219&view=results and the first failing build was https://dev.azure.com/dnceng/internal/_build/results?buildId=2359821&view=results.

The runtime range of commits between these two is https://github.com/dotnet/runtime/compare/8accd8073a44f5835b39e0584f17d1aff1771121...e99836a899bdb97be5afbbed8d90573ee5635b30

The commits that pop out to me as possibly causing crossgen2 crashes are:

- https://github.com/dotnet/runtime/commit/205ef031e0fe5152dede0bd9f99d0f6f9e7f1e45
- https://github.com/dotnet/runtime/commit/d00587db256d5684ba7458225c35a06347b487eb

I think the second is more likely than the first.

cc: @EgorBo

**Comment by EgorBo** [comment]
Checking now

**Comment by mthalman** [comment]
It's getting a different error now and failing with error code 134:

```
/vmr/src/installer/src/redist/targets/Crossgen.targets(172,5): *** stack smashing detected ***: terminated [/vmr/src/installer/src/redist/redist.csproj]
```

[binlog](https://github.com/dotnet/source-build/files/14039103/sourcebuild.installer.2.zip)


[Failing build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2360900&view=logs&j=0bf36158-843f-57d9-fe78-bca5fcea110c&t=48dffdfd-1582-5ea1-3d8d-b9635049aefa) (internal link)

**Comment by mthalman** [comment]
It's consistency failing but not always failing with the same error. The two errors mentioned in this issue above are all I've seen so far.

**Comment by mthalman** [comment]
Interestingly, the build succeeded in this [recent run](https://dev.azure.com/dnceng/internal/_build/results?buildId=2361439&view=results) (internal link). But I'm not sure I trust that the issue is resolved. Is it perhaps an intermittent issue? The source changes that exist between the two builds don't include any runtime changes at all.

**Comment by jkoritzinsky** [comment]
Yes, it is possible that the failure is intermittent given your sample with the 134 exit code. It's likely a native corruption somewhere in crossgen2. If this is using a live crossgen2 (which I think it is), then it could be anywhere in the JIT, GC, or CoreCLR runtime. This also means it could have appeared before the first instance during the time frame where the build was failing before getting to installer, which makes this harder to diagnose...

**Comment by mthalman** [comment]
There's been 3 other builds since that one that passed and they've all failed.

**Comment by agocke** [comment]
@MichaelSimons is there an easy way to test a build with a runtime change reverted? We might see some benefit at just looking at a build with each of those two commits that Jeremy identified reverted

**Comment by BruceForstall** [comment]
@mthalman I see you're testing a revert of @EgorBo 's change here: https://dev.azure.com/dnceng/internal/_build?definitionId=1219. Could you also test a revert of https://github.com/dotnet/runtime/commit/205ef031e0fe5152dede0bd9f99d0f6f9e7f1e45 in parallel just to be comprehensive in our checking? (It seems unlikely to be the problem; it should have only deleted unused code)

**Comment by BruceForstall** [comment]
Oh, I guess @EgorBo triggered those runs; @EgorBo maybe you could trigger runs with Jakob's change reverted?

**Comment by EgorBo** [comment]
So current status: I am pretty sure it's not https://github.com/dotnet/runtime/commit/d00587db256d5684ba7458225c35a06347b487eb although it looked like it was because several runs with the revert were green but it turned out the error doesn't always manifest.

I've reproduced it locally but when it failed and I extracted the exact crossgen command and ran it again:
```
/home/egorbo/dotnet-dotnet/src/installer/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.ubuntu.20.04-arm64/9.0.0-preview.1.24073.11/tools/crossgen2 \
"/home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-preview.1.24075.1/FSharp/FSharp.Compiler.Service.dll" \
-o "/tmp/bxhyzqnu.o3h/FSharp.Compiler.Service.dll" \
--targetarch arm64 \
-r /home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-preview.1.24075.1/FSharp/runtimes/win/lib/net9.0/*.dll \
-r /home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-preview.1.24075.1/FSharp/*.dll \
-r /home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/shared/Microsoft.NETCore.App/9.0.0-preview.1.24073.11/*.dll
```
it failed with:
```
Unhandled exception. System.CommandLine.CommandLineException: Multiple input files matching same simple name /home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-preview.1.24075.1/FSharp/FSharp.Compiler.Service.dll /home/egorbo/dotnet-dotnet/src/installer/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-preview.1.24075.1/FSharp/FSharp.Compiler.Service.dll
   at System.CommandLine.Helpers.AppendExpandedPaths(Dictionary`2, String, Boolean)
   at System.CommandLine.Helpers.BuildPathDictionary(IReadOnlyList`1, Boolean)
   at System.CommandLine.CliArgument`1.<>c__DisplayClass8_0.<set_CustomParser>b__0(ArgumentResult argumentResult, Object& parsedValue)
   at System.CommandLine.Parsing.ArgumentResult.ValidateAndConvert(Boolean)
   at System.CommandLine.Parsing.CommandResult.ValidateArguments(Boolean)
   at System.CommandLine.Parsing.ParseOperation.Validate()
   at System.CommandLine.Parsing.ParseOperation.Parse()
   at System.CommandLine.Parsing.CliParser.Parse(CliCommand, IReadOnlyList`1, String , CliConfiguration )
   at System.CommandLine.CliConfiguration.Invoke(String[])
   at ILCompiler.Program.Main(String[])
run.sh: line 7: 1979606 Aborted                 (core dumped)
```
with error code 134.

So it could be some high-level issue in crossgen/its config or it's still could be a codegen bug that changed crossgen's behaviour?

**Comment by oleksandr-didyk** [comment]
The latest `main` VMR is red on a crossgen failure that I didn't find in this issue, so linking it in case its related or might help ([source](https://dev.azure.com/dnceng/internal/_build/results?buildId=2366431&view=logs&j=0bf36158-843f-57d9-fe78-bca5fcea110c&t=48dffdfd-1582-5ea1-3d8d-b9635049aefa))

```
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB6006: "crossgen2" exited with code 139. [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018: The "Crossgen" task failed unexpectedly. [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018: System.IO.IOException: Directory not empty : '/tmp/xinmgcn5.ilw' [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018:    at System.IO.FileSystem.RemoveEmptyDirectory(String fullPath, Boolean topLevel, Boolean throwWhenNotEmpty) [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Crossgen.Execute() in /_/src/installer/src/core-sdk-tasks/Crossgen.cs:line 81 [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/vmr/src/installer/src/redist/redist.csproj]
    /vmr/src/installer/src/redist/targets/Crossgen.targets(181,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/vmr/src/installer/src/redist/redist.csproj]

```

**Comment by EgorBo** [comment]
Long story short: the issue was handled offline and it turned out that the issue was introduced in https://github.com/dotnet/runtime/pull/96969 and the fix (revert) is https://github.com/dotnet/runtime/pull/97679 (although, it's probably reverted only from release/9.0p1 branch?)

**Comment by oleksandr-didyk** [comment]
> although, it's probably reverted only from release/9.0p1 branch?

Yes, and the latest VMR builds for `release/9.0` are green ([source](https://dev.azure.com/dnceng/internal/_build?definitionId=1219&_a=summary&branchFilter=256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092)).

`main` is still failing in the same Ubuntu leg as before with the exception linked above. Since the original issue was about `preview-1` and the exceptions seem to be different, should we open a different one for tracking the failure in `main`?

**Comment by mthalman** [comment]
> > although, it's probably reverted only from release/9.0p1 branch?
> 
> Yes, and the latest VMR builds for `release/9.0` are green ([source](https://dev.azure.com/dnceng/internal/_build?definitionId=1219&_a=summary&branchFilter=256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092%2C256092)).
> 
> `main` is still failing in the same Ubuntu leg as before with the exception linked above. Since the original issue was about `preview-1` and the exceptions seem to be different, should we open a different one for tracking the failure in `main`?

It's not a separate issue. The fix wasn't applied to `main`. See https://github.com/dotnet/runtime/pull/97679#issuecomment-1921182993

**Comment by mthalman** [comment]
Should be fixed by https://github.com/dotnet/runtime/pull/97817

**Comment by mthalman** [comment]
I'm not sure that the changes in https://github.com/dotnet/runtime/pull/97817 fixed the issue. Either that or there's another related issue. This change finally flowed into the VMR on Friday and the build right after it came in failed in the Arm64 build leg when calling crossgen2. This time it was during the `_CreateR2RImages` target. [Related build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2375141&view=logs&j=8e1d7f5f-2dd9-5bba-3568-6eebce65bc74&t=a8f9fb18-43e6-5d18-cb2e-d8cac8cb066e) (internal link). The build after it passed however. But we previously saw inconsistent results for this issue as well. I'm running a full build again to get a another data point.

**Comment by mthalman** [comment]
The new build passed as well. I'm going to close this and hope things are ok. I'll reopen if it pops up again.

**Comment by ellahathaway** [comment]
@mthalman - I ran from the tip around noon today and encountered the error. [Build link](https://dev.azure.com/dnceng/internal/_build/results?buildId=2376696&view=logs&j=e141397f-257a-5c76-d304-98fe1c061b34&t=eaf2c6c8-dabf-54a6-858b-22e2b1087234&l=6948) (internal). It's possible that the change hadn't flowed in yet (parent commit was https://dev.azure.com/dnceng/internal/_git/dotnet-dotnet/commit/b07c6879698f9df26ae94f8d9c3e63b721242207?refName=refs%2Fheads%2Fdev%2Fellahathaway%2Ftest), but I wanted to make you aware.

**Comment by mthalman** [comment]
Yep, it's still happening and is intermittent. Pinging @janvorli since the changes in https://github.com/dotnet/runtime/pull/97817 don't seem to be sufficient. It's gotten farther in the build than before but now fails when invoking crossgen2 as part of creating R2R images. Again, this is still limited to the Arm64 build leg.

```
/vmr/src/runtime/artifacts/bin/Crossgen2Tasks/Release/net9.0/Microsoft.NET.CrossGen.targets(469,5): error MSB6006: "/vmr/.dotnet/dotnet" exited with code 139. [/vmr/src/runtime/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj]
    OutputProperties
        _ReadyToRunCompilerExitCode = 139
        _ReadyToRunWarningsDetected = False
```

[pipeline run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=564091&view=results)
[binlog](https://artprodcus3.artifacts.visualstudio.com/A6fcc92e5-73a7-4f88-8d13-d9045b45fb27/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2RuY2VuZy1wdWJsaWMvcHJvamVjdElkL2NiYjE4MjYxLWM0OGYtNGFiYi04NjUxLThjZGNiNTQ3NDY0OS9idWlsZElkLzU2NDA5MS9hcnRpZmFjdE5hbWUvVWJ1bnR1MjIwNEFybTY0X09mZmxpbmVfTXNmdFNka19hcm02NF9CdWlsZExvZ3NfQXR0ZW1wdDE1/content?format=file&subPath=%2Fsrc%2Fruntime%2Fartifacts%2Flog%2FRelease%2Fsource-inner-build.binlog)

**Comment by janvorli** [comment]
@mthalman can you please share repro instructions with me? I am not familiar with the source build and while I have tried to mimic what I've seen in the ci log, I am not sure if I did it right.
I am building it on a physical arm64 linux device with Ubuntu 22.04, so I am not using docker. The commands I've tried to use were:
```sh
./prep.sh
BUILD_SOURCEVERSION=220c455bc727ab0a8bdd9feb61f474a605d30339 ./build.sh --ci --clean-while-building --prepareMachine --source-only
```
(I have used the current main for build, so I've changed the 220c455bc727ab0a8bdd9feb61f474a605d30339 to the commit hash of my checkout)

**Comment by mthalman** [comment]
It should be enough to just run these commands:

```
./prep.sh
./build.sh --sb
```

Depending on how much disk space you have available, you may need to run `--clean-while-building` which cleans up repo outputs after they're done building successfully. That cleans up binaries but leaves logs.

**Comment by janvorli** [comment]
@mthalman my build keeps failing due to this:
CSC : error CS1566: Error reading resource '_framework/blazor.web.js' -- 'Could not find a part of the path '/home/janvorli/git/dotnet/src/aspnetcore/src/Components/Web.JS/dist/Release/blazor.web.js'.' [/home/janvorli/git/dotnet/src/aspnetcore/src/Components/Endpoints/src/Microsoft.AspNetCore.Components.Endpoints.csproj] [/home/janvorli/git/dotnet/repo-projects/aspnetcore.proj]

I've tried several times, even with git clean -xdf in between. Any idea what is wrong?

**Comment by mthalman** [comment]
That looks related to not having the latest source. Have you pulled latest from main branch of the VMR? That error was relevant in an older commit but was fixed by https://github.com/dotnet/installer/pull/18641/commits/90ccc0ca4d42e91029a392fa210fdca06447b50b in https://github.com/dotnet/installer/pull/18641.

Oh, and you'll also need to run it in with the `--online` switch to avoid https://github.com/dotnet/source-build/issues/4129.

**Comment by janvorli** [comment]
I did a fresh clone today, I didn't have that repo locally. I am on this commit:
```
commit dc4df39f794e6b9c565d721986b71a056288b812 (HEAD -> main, origin/main, origin/HEAD)
Merge: 55519d6838 85d7a9e37c
Author: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Date:   Fri Feb 16 13:18:50 2024 +0000

    [Recursive sync] installer / 6d1abb7 → 9b54914

    Updated repositories:
      - installer / 6d1abb7 → 9b54914
        https://github.com/dotnet/installer/compare/6d1abb74140d80b81d4f482a49332d55f89468d7..9b54914494e95df2d6e72803dfe7d2193cdebb22
      - aspire / 8ec92cb → b2107ab
        https://github.com/dotnet/aspire/compare/8ec92cbc5fbcba7a677fb52aaa4b0118f1ed17f4..b2107ab9a2b96b609f16997525f694f587e3018e

    [[ commit created by automation ]]

```

**Comment by janvorli** [comment]
I've synced to the latest state now again and I am retrying.

**Comment by mthalman** [comment]
Interesting. Well that might be another issue with Arm then. Since this is failing in the build of the aspnetcore repo, it means you passed the point of the crossgen failure in the runtime repo (runtime builds before aspnetcore). So this means you're not repro'ing the issue. Be aware that this is an intermittent issue so you may need to build several times to get it. To get a fresh VMR to start over from, I execute `git clean -fdx` and `git checkout -- .`

**Comment by janvorli** [comment]
Ok, I'll ignore that error then and just retry a clean build.

**Comment by agocke** [comment]
I don't think this has anything to do with R2R itself. The failure in the linked issue is due to an unhandled exception. The error message is:

```
      Unhandled exception. System.CommandLine.CommandLineException: Target OS 'centos' is not supported
         at System.CommandLine.Helpers.GetTargetOS(String) in /_/src/runtime/src/coreclr/tools/Common/CommandLineHelpers.cs:line 84
         at System.CommandLine.CliArgument`1.<>c__DisplayClass8_0.<set_CustomParser>b__0(ArgumentResult argumentResult, Object& parsedValue)
         at System.CommandLine.Parsing.ArgumentResult.ValidateAndConvert(Boolean)
         at System.CommandLine.Parsing.CommandResult.ValidateOptions(Boolean)
         at System.CommandLine.Parsing.CommandResult.Validate(Boolean)
         at System.CommandLine.Parsing.ParseOperation.Validate()
         at System.CommandLine.Parsing.ParseOperation.Parse()
         at System.CommandLine.Parsing.CliParser.Parse(CliCommand, IReadOnlyList`1, String , CliConfiguration )
         at System.CommandLine.CliConfiguration.Invoke(String[])
         at ILCompiler.Program.Main(String[]) in /_/src/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 753
```

**Comment by janvorli** [comment]
@agocke the Ubuntu run seems to be failing with 139 during crossgen2 run though.

**Comment by agocke** [comment]
Ah, you're right I was lookihg at a different log

**Comment by mthalman** [comment]
> I don't think this has anything to do with R2R itself. The failure in the linked issue is due to an unhandled exception. The error message is:
> 
> ```
>       Unhandled exception. System.CommandLine.CommandLineException: Target OS 'centos' is not supported
>          at System.CommandLine.Helpers.GetTargetOS(String) in /_/src/runtime/src/coreclr/tools/Common/CommandLineHelpers.cs:line 84
>          at System.CommandLine.CliArgument`1.<>c__DisplayClass8_0.<set_CustomParser>b__0(ArgumentResult argumentResult, Object& parsedValue)
>          at System.CommandLine.Parsing.ArgumentResult.ValidateAndConvert(Boolean)
>          at System.CommandLine.Parsing.CommandResult.ValidateOptions(Boolean)
>          at System.CommandLine.Parsing.CommandResult.Validate(Boolean)
>          at System.CommandLine.Parsing.ParseOperation.Validate()
>          at System.CommandLine.Parsing.ParseOperation.Parse()
>          at System.CommandLine.Parsing.CliParser.Parse(CliCommand, IReadOnlyList`1, String , CliConfiguration )
>          at System.CommandLine.CliConfiguration.Invoke(String[])
>          at ILCompiler.Program.Main(String[]) in /_/src/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 753
> ```

That was a separate issue (identified in https://github.com/dotnet/source-build/issues/4111#issuecomment-1938755018), that has since been fixed.

**Comment by janvorli** [comment]
@mthalman mystery solved. It is not a new issue. The thing is that crossgen doesn't use the current build of runtime, but rather an older version (it uses the dotnet sdk in the .dotnet folder). In this case, it used version 9.0.0-preview.2.24080.1 built from commit d40c654c274fe4f4afe66328f0599130f3eb2ea6 that predates the fix in main by two days.

**Comment by mthalman** [comment]
Ok, I'll update the VMR to use a more recent SDK and check the results.

**Comment by ellahathaway** [comment]
This is continuing to be an issue. The most recent related error that I've seen is in runtime:

`/vmr/src/runtime/artifacts/bin/Crossgen2Tasks/Release/net9.0/Microsoft.NET.CrossGen.targets(357,5): error NETSDK1096: Optimizing assemblies for performance failed. You can either exclude the failing assemblies from being optimized, or set the PublishReadyToRun property to false. [/vmr/src/runtime/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.sfxproj]`

[Link to build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2392771&view=logs&j=8e1d7f5f-2dd9-5bba-3568-6eebce65bc74&t=a8f9fb18-43e6-5d18-cb2e-d8cac8cb066e&l=6927) (internal link)

**Comment by mthalman** [comment]
The bootstrap work to resolve this is still pending.

**Comment by mthalman** [comment]
This should be fixed by https://github.com/dotnet/installer/pull/18763

**Comment by sec** [comment]
Hi,
I also started to hit this issue, just under FreeBSD, could I ask for some info what was the issue and/or which commit fixed that?

**Comment by mthalman** [comment]
Reopening this because it's not yet resolved for source build in the VMR. We're still waiting for a bootstrap of the VMR onto an SDK build which contains the fix for this issue. But that's currently blocked on https://github.com/dotnet/source-build/issues/4206.

**Comment by mthalman** [comment]
> Hi, I also started to hit this issue, just under FreeBSD, could I ask for some info what was the issue and/or which commit fixed that?

@sec - This is the fix: https://github.com/dotnet/runtime/pull/97817. But it's not enough to have the fix in the source that you're currently building. You need to build with an SDK that has this fix. That's what we're currently working towards for the [VMR](https://github.com/dotnet/dotnet), to get it bootstrapped with an SDK that has this fix in it. That will allow builds to work using that SDK.

**Comment by sec** [comment]
@mthalman Thanks for the info. If I get it correctly, I should wait for your work to be done, then I can crosscompile SDK for FreeBSD and use it as bootstrap for preview-2, which should be fine. BUT, as looking at this, it's a bug in libunwind - which under FreeBSD is taken from ports - looking in there I've already found https://github.com/libunwind/libunwind/issues/715 - this should also be fixed otherwise I will still hit this in future :)

**Comment by mthalman** [comment]
The fix was flowed into the VMR with https://github.com/dotnet/installer/pull/19145
